SwiftCompilerSources/Sources/Optimizer/FunctionPasses/EffectInference.swift

```diff
@@ -0,0 +1,197 @@
+//===--- EffectInference.swift - Computes side-effects          ----------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+let effectInference = FunctionPass(name: "infer-effects", {
+  (function: Function, context: PassContext) in
+  var effectState = EffectState(function.entryBlock.arguments.endIndex, context.calleeAnalysis)
+
+  for block in function.blocks {
+    for inst in block.instructions {
+      effectState.update(for: inst)
+    }
+  }
+
+  context.modifyEffects(in: function, { effects in
+    effects.globalEffects = effectState.globalEffects
+    for (idx, (effect, path)) in effectState.argumentEffects.enumerated() {
+      effects.argumentEffects.append(
+        ArgumentEffect(
+          .sideeffect(effect),
+          selectedArg: ArgumentEffect.Selection(.argument(idx), pathPattern: effect.isPure ? SmallProjectionPath(): path!)
+        )
+      )
+    }
+  })
+})
+
+
+private struct EffectState {
+  var argumentEffects: [(effect: SideEffect, path: SmallProjectionPath?)]
+  var globalEffects: GlobalEffects = GlobalEffects(isValid: true)
+  var localEffects: SideEffect = SideEffect()
+
+  var accessPathWalker = AccessPathWalker()
+  var accessStorageWalker = AccessStoragePathWalker()
+
+  private let calleeAnalysis: CalleeAnalysis
+
+  init(_ nargs: Int, _ calleeAnalysis: CalleeAnalysis) {
+    argumentEffects = Array(repeating: (SideEffect(), nil), count: nargs)
+    self.calleeAnalysis = calleeAnalysis
+  }
+
+  mutating
+  func update(for inst: Instruction) {
+    var matched = true
+    switch inst {
+    case let apply as ApplyInst:
+      guard let callees = calleeAnalysis.getCallees(callee: apply.callee) else { return globalEffects.setWorstEffects() }
+
+      for callee in callees {
+        // Collect callee `sideeffect`s
+        let effects = callee.effects
+        let sideeffects = effects.argumentEffects.lazy.compactMap({
```

Comment:
Why doing a `compactMap` here?
And not just
```
for effect in effects.argumentEffects {
// loop body from the "Update argument effects" loop below
}
```

Reply:
Here we are collecting only the `sideeffect` effects, excluding `escape` effects.

Reply:
Sure, I meant
```
for effect in effects.argumentEffects {
  if effect.kind == .sideEffect {
    ...
  }
}
```